src/frontend/src/optimizer/mod.rs

```diff
@@ -576,11 +601,7 @@ impl PlanRoot {
             ctx.trace("To Batch Distributed Plan:");
             ctx.trace(plan.explain_to_string().unwrap());
         }
-        if has_batch_insert(plan.clone())
-            || has_batch_delete(plan.clone())
-            || has_batch_update(plan.clone())
-            || Self::require_additional_exchange_on_root(plan.clone())
-        {
+        if Self::require_additional_exchange_on_root(plan.clone()) {
```

Comment:
`require_additional_exchange_on_root` now will find the insert, delete, update no exchange before so I think we can remove has_** here. 

Although the behaviour of `has_**` and `exist_and_no_exchange_before` is different, I think it's still correct now? CMIIW.